src/puppetlabs/puppetdb/utils/string_formatter.clj

```diff
@@ -0,0 +1,101 @@
+                 ["\\E"])))
+
+(defn maybe-strip-escaped-quotes
+  [s]
+  (if (and (> (count s) 1)
+           (string/starts-with? s "\"")
+           (string/ends-with? s "\""))
+    (subs s 1 (dec (count s)))
+    s))
+
+(defn quoted
+  [s]
+  (str "'" s "'"))
+
+(defn comma-separated-keywords
+  [words]
+  (let [quoted-words (map quoted words)]
+    (if (> (count quoted-words) 2)
+      (str (string/join ", " (butlast quoted-words)) ", " "and " (last quoted-words))
+      (string/join " and " quoted-words))))
+
+(defn dashes->underscores
+  "Accepts a string or a keyword as an argument, replaces all occurrences of the
+  dash/hyphen character with an underscore, and returns the same type (string
+  or keyword) that was passed in.  This is useful for translating data structures
+  from their wire format to the format that is needed for JDBC."
+  [str]
+  (let [result (string/replace (name str) \- \_)]
+    (if (keyword? str)
+      (keyword result)
+      result)))
+
+(defn underscores->dashes
+  "Accepts a string or a keyword as an argument, replaces all occurrences of the
+   underscore character with a dash, and returns the same type (string
+   or keyword) that was passed in.  This is useful for translating data structures
+   from their JDBC-compatible representation to their wire format representation."
+  [s]
+  (let [opt-key? (optional-key? s)
+        result (if opt-key?
+                 (string/replace (name (:k s)) \_ \-)
+                 (string/replace (name s) \_ \-))]
+    (cond
+      opt-key? (if (keyword? (:k s))
+                 (s/optional-key (keyword result))
+                 (s/optional-key result))
+      (keyword? s) (keyword result)
+      :else result)))
+
+(defn dash->underscore-keys
+  "Converts all top-level keys (including nested maps) in `m` to use dashes
+  instead of underscores as word separators"
+  [m]
+  (sp/transform [sp/ALL]
+                #(update % 0 dashes->underscores)
+                m))
+
+(defn underscore->dash-keys
+  "Converts all top-level keys (including nested maps) in `m` to use underscores
+  instead of underscores as word separators"
+  [m]
+  (sp/transform [sp/ALL]
+                #(update % 0 underscores->dashes)
+                m))
+
+(defn pprint-json-parse-exception
+  "Returns a parsed JsonParseException message.
+  From the second line of the error message, the line and
+  column index are extracted, so that we can place an arrow,
+  pointing at that position. "
+  [error query]
+  (let [error-lines (string/split-lines (.getMessage error))
+        [_ line column] (re-find #"line: (\d+), column: (\d+)" (second error-lines))
+        line-index (Integer. line)
+        column-index (Integer. column)
+        ;subtracting 1 from the line and column index,
+        ;because they start from 1, not 0
+        query-line-index (dec line-index)
+        query-column-index (dec column-index)]
+
+    (str "Json parse error at line " line-index ", column " column-index ":\n\n"
```

Comment:
@Zak-Kent are we using i18n consistently in PDB?  If so, these should likely be using tru (trs?) : https://github.com/puppetlabs/clj-i18n#developer-usage

Reply:
Good catch, I wasn't even thinking about the i18n stuff. It looks like the the result of this func is wrapped in a `tru` in this PR but we could move that wrapping inside this func so we don't need to remember it in the future.

Reply:
I think if we wrap that top line we're creating here in a `tru` that might be what we want. It would look like this: 
```(str (tru "Json parse error at line {0}, column {1}:\n\n" line-index column-index) ...)``` I think we can get away with not translating the rest because it's coming from the query the user provided and the stack trace from the JsonParseException.